btree/btree.c

```diff
@@ -2377,6 +2411,288 @@ void m0_btree_ut_node_add_del_rec(void)
+
+	ksize = m0_vec_count(&datum->key->k_data.ov_vec);
+	M0_PRE(m0_vec_count(&rec->r_key.k_data.ov_vec) <= ksize);
+
+	vsize = m0_vec_count(&datum->value->ov_vec);
+	M0_PRE(m0_vec_count(&rec->r_val.ov_vec) <= vsize);
+
+	m0_bufvec_cursor_init(&dcur, &datum->key->k_data);
+	m0_bufvec_cursor_init(&scur, &rec->r_key.k_data);
+	m0_bufvec_cursor_copy(&dcur, &scur, ksize);
+
+	m0_bufvec_cursor_init(&dcur, datum->value);
+	m0_bufvec_cursor_init(&scur, &rec->r_val);
+	m0_bufvec_cursor_copy(&dcur, &scur, vsize);
+
+	return 0;
+}
+
+
+/**
+ * This unit test exercises the KV operations for both valid and invalid
+ * conditions.
+ */
+static void m0_btree_ut_basic_kv_operations(void)
+{
+	struct m0_btree_type  btree_type   = {.tt_id = M0_BT_EMAP_EM_MAPPING};
+	struct m0_be_tx      *tx           = NULL;
+	struct m0_btree_op    b_op;
+	void                 *temp_node;
+	int                   i;
+	time_t                curr_time;
+	struct m0_btree_cb    ut_cb;
+	uint64_t              first_key;
+	bool                  first_key_initialized = false;
+	struct m0_btree_op    kv_op;
+
+	M0_ENTRY();
+
+	time(&curr_time);
+	printf("\nUsing seed %lu", curr_time);
+	srandom(curr_time);
+
+	/** Prepare transaction to capture tree operations. */
+	m0_be_tx_init(tx, 0, NULL, NULL, NULL, NULL, NULL, NULL);
+	m0_be_tx_prep(tx, NULL);
+
+	/**
+	 *  Run valid scenario:
+	 *  1) Create a btree
+	 *  2) Adds a few records to the created tree.
+	 *  3) Confirms the records are present in the tree.
+	 *  4) Deletes all the records from the tree.
+	 *  4) Close the btree
+	 *  5) Destroy the btree
+	 */
+
+	/** Create temp node space and use it as root node for btree */
+	temp_node = m0_alloc_aligned((1024 + sizeof(struct nd)), 10);
+	M0_BTREE_OP_SYNC_WITH(&b_op.bo_op,
+			      m0_btree_create(temp_node, 1024, &btree_type,
+					      tx, &b_op));
+
+	for (i = 0; i < 2048; i++) {
+		uint64_t             key;
+		uint64_t             value;
+		struct cb_data       put_data;
+		struct m0_btree_key  put_key;
+		struct m0_bufvec     put_value;
+		m0_bcount_t          ksize  = sizeof key;
+		m0_bcount_t          vsize  = sizeof value;
+		void                *k_ptr  = &key;
+		void                *v_ptr  = &value;
+
+		/**
+		 *  There is a very low possibility of hitting the same key
+		 *  again. This is fine as it helps debug the code when insert
+		 *  is called with the same key instead of update function.
+		 */
+		key = value = random();
+
+		if (!first_key_initialized) {
+			first_key = key;
+			first_key_initialized = true;
+		}
+
+		put_key.k_data     = M0_BUFVEC_INIT_BUF(&k_ptr, &ksize);
+		put_value          = M0_BUFVEC_INIT_BUF(&v_ptr, &vsize);
+
+		put_data.key       = &put_key;
+		put_data.value     = &put_value;
+
+		ut_cb.c_act        = btree_kv_put_cb;
+		ut_cb.c_datum      = &put_data;
+
+		M0_BTREE_OP_SYNC_WITH(&kv_op.bo_op,
+				      m0_btree_put(b_op.bo_arbor, tx,
+						   &put_key, &ut_cb, 0,
+						   &kv_op));
+	}
+
+	{
```

Comment:
Bracket and new variables at the middle of function looks a bit weird. Can't the local variables be defined at the beginning of the function and also some of the already used local variable be reused ?

Reply:
This makes the code separate like the upper code section covered the record addition while this section covers the key searches.